video/out/drm_prime.c

```diff
@@ -78,15 +81,71 @@ int drm_prime_create_framebuffer(struct mp_log *log, int fd, AVDRMFrameDescripto
    return -1;
 }
 
-void drm_prime_destroy_framebuffer(struct mp_log *log, int fd, struct  drm_prime_framebuffer *framebuffer)
+void drm_prime_destroy_framebuffer(struct mp_log *log, int fd,
+                                    struct drm_prime_framebuffer *framebuffer,
+                                    struct drm_prime_handle_refs *handle_refs)
 {
     if (framebuffer->fb_id)
         drmModeRmFB(fd, framebuffer->fb_id);
 
     for (int i = 0; i < AV_DRM_MAX_PLANES; i++) {
-        if (framebuffer->gem_handles[i])
-            drmIoctl(fd, DRM_IOCTL_GEM_CLOSE, &framebuffer->gem_handles[i]);
+        if (framebuffer->gem_handles[i]) {
+            drm_prime_remove_handle_ref(handle_refs,
+                                         framebuffer->gem_handles[i]);
+            if(!drm_prime_get_handle_ref_count(handle_refs, framebuffer->gem_handles[i])) {
+                drmIoctl(fd, DRM_IOCTL_GEM_CLOSE, &framebuffer->gem_handles[i]);
+            }
+        }
     }
 
     memset(framebuffer, 0, sizeof(*framebuffer));
 }
+
+void drm_prime_init_handle_ref_count(struct drm_prime_handle_refs *handle_refs) {
+    //4 dumb buffer + (1 per plane * planes we keep) + 10 (extra safety)
+    uint16_t init_size = 4 + (AV_DRM_MAX_PLANES * 4) + 10;
+    handle_refs->handle_ref_count = (uint32_t *)malloc(init_size * sizeof(uint32_t));
+    handle_refs->size = init_size;
+
+    memset(handle_refs->handle_ref_count, 0, sizeof(uint32_t) * init_size);
+}
+
+void drm_prime_add_handle_ref(struct drm_prime_handle_refs *handle_refs, uint32_t handle) {
+    if(handle) {
+        if(handle > handle_refs->size) {
+            handle_refs->size *= 2;
+            handle_refs->handle_ref_count =
+             (uint32_t *)realloc(handle_refs->handle_ref_count,
+                                  handle_refs->size * sizeof(uint32_t));
```

Comment:
This doesn't check for errors, overflows, and the cast is unnecessary.
You may be able to use `MP_TARRAY_GROW`, but that requires using the ta/talloc functions instead maloc/realloc/free.

Reply:
I'm trying to get it working with talloc but I've never used that before so need to get known to it. But doesn't seem to be very hard.

Reply:
See the latest commit. I hpe I used it correctly.